geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/AvroSimpleFeatureTypeParser.scala

```diff
@@ -0,0 +1,458 @@
+          srid.map(builder.get(fieldName).getUserData.put(SimpleFeatureTypes.AttributeOptions.OptSrid, _))
+        case DateProperties(_, default) =>
+          builder.add(fieldName, classOf[Date])
+          // if more than one field is set as the default date, the last one becomes the default
+          if (default) { userData.put(SimpleFeatureTypes.Configs.DefaultDtgField, fieldName) }
+        case VisibilityProperties =>
+          // if there is more than one visibilities field, the last one is used
+          userData.put(GeomesaAvroFeatureVisibility.KEY, fieldName)
+          addFieldToBuilder(builder, field)
+        case NoProperties =>
+          addFieldToBuilder(builder, field)
+      }
+
+      // any extra props on the field go in the attribute user data
+      builder.get(fieldName).getUserData.putAll(metadata.extra)
+    }
+
+    val sft = builder.buildFeatureType()
+    sft.getUserData.putAll(userData)
+    sft
+  }
+
+  @tailrec
+  private def addFieldToBuilder(builder: SimpleFeatureTypeBuilder,
+                                field: Schema.Field,
+                                typeOverride: Option[Schema.Type] = None): Unit = {
+    typeOverride.getOrElse(field.schema().getType) match {
+      case Schema.Type.STRING  => builder.add(field.name(), classOf[java.lang.String])
+      case Schema.Type.BOOLEAN => builder.add(field.name(), classOf[java.lang.Boolean])
+      case Schema.Type.INT     => builder.add(field.name(), classOf[java.lang.Integer])
+      case Schema.Type.DOUBLE  => builder.add(field.name(), classOf[java.lang.Double])
+      case Schema.Type.LONG    => builder.add(field.name(), classOf[java.lang.Long])
+      case Schema.Type.FLOAT   => builder.add(field.name(), classOf[java.lang.Float])
+      case Schema.Type.BYTES   => builder.add(field.name(), classOf[Array[Byte]])
+      case Schema.Type.UNION   =>
+        // if a union has more than one non-null type, it is not supported
+        val types = field.schema().getTypes.map(_.getType).filter(_ != Schema.Type.NULL).toSet
+        if (types.size != 1) {
+          throw UnsupportedAvroTypeException(types.mkString("[", ", ", "]"))
+        } else {
+          addFieldToBuilder(builder, field, Option(types.head))
+        }
+      case Schema.Type.MAP     => throw UnsupportedAvroTypeException(Schema.Type.MAP.getName)
+      case Schema.Type.RECORD  => throw UnsupportedAvroTypeException(Schema.Type.RECORD.getName)
+      case Schema.Type.ENUM    => builder.add(field.name(), classOf[java.lang.String])
+      case Schema.Type.ARRAY   => throw UnsupportedAvroTypeException(Schema.Type.ARRAY.getName)
+      case Schema.Type.FIXED   => throw UnsupportedAvroTypeException(Schema.Type.FIXED.getName)
+      case Schema.Type.NULL    => throw UnsupportedAvroTypeException(Schema.Type.NULL.getName)
+      case _                   => throw UnsupportedAvroTypeException("unknown")
+    }
+  }
+
+  private def parseMetadata(field: Schema.Field): GeomesaAvroFieldMetadata = {
```

Comment:
I'm picky about the M in GeoMesa being capitalized.  The refactor in IntelliJ should make that quick.

Reply:
Will do